### PR TITLE
Add `DayPickerSingleDatePicker` component

### PR DESCRIPTION
**Pull request by majapw**
Fix for https://github.com/airbnb/react-dates/issues/524

I've created a `DayPickerSingleDateController` to match the `DayPickerRangeController` component. I've also added stories for it as well.

The related PR for ranges is https://github.com/airbnb/react-dates/pull/167. The basic gist of this work is that all calendar related code has been moved into an abstraction layer in between the `DayPicker` and the `SingleDatePicker` called the `DayPickerSingleDateController`. Specifically this includes the day interaction callbacks (`onDayMouseEnter`, `onDayMouseLeave`, and `onDayClick`) and all code relating to the modifiers. 

The following graphic should help explain the new structure somewhat (`DayPickerWithRangeControllers` maps to the `DayPickerSingleDateController`).
new SDP structure

In a future PR, I will separate out the code related strictly to the input from the `SingleDatePicker`. After that the `SingleDatePicker` will strictly contain code that coordinates focus and visibility between the input and the calendar dropdown.

to: @tagoro9 @pokonski @airbnb/webinfra

### CODE REVIEW

**Comment by pokonski** [comment]
Fantastic news! I'll check it out very soon, thanks Maja 👏

**Comment by tagoro9** [comment]
This is awesome! Thanks so much!

**Comment by majapw** [comment]
@ljharb I've added a lot more content to the description to provide context for the changes if you would like to take a look! It does largely just entail moving around a functions that already existed and I do have confidence in the tests catching things. :D

**Comment by tagoro9** [comment]
I just switched to this version in our library and replaced the uses of our custom implementation of `DayPickerSingleDateController` with the one in here and everything is working great and test are passing. This is awesome! Thank you very much!

I just have one question. `DayPickerSingleDateController` is not exported in the `index.js` file. Are you planning on exporting it?

File: src/components/DayPickerSingleDateController.jsx [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+    this.onDayClick = this.onDayClick.bind(this);
+
+    this.onPrevMonthClick = this.onPrevMonthClick.bind(this);
+    this.onNextMonthClick = this.onNextMonthClick.bind(this);
+
+    this.getFirstFocusableDay = this.getFirstFocusableDay.bind(this);
+  }
+
+  componentDidMount() {
+    this.isTouchDevice = isTouchDevice();
```
Comment (ljharb) [inline src/components/DayPickerSingleDateController.jsx]: should this trigger a rerender, if it changes?

File: src/components/DayPickerSingleDateController.jsx [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+      visibleDays[month].forEach((day) => {
+        modifiers[month][toISODateString(day)] = this.getModifiersForDay(day);
+      });
+    });
+
+    return modifiers;
+  }
+
+  getModifiersForDay(day) {
+    return new Set(Object.keys(this.modifiers).filter(modifier => this.modifiers[modifier](day)));
```
Comment (ljharb) [inline src/components/DayPickerSingleDateController.jsx]: should this be `new Set(Object.values(this.modifiers).map(x => x(day)))`? (using `object.values` ofc)

(fine as-is, too)

File: src/components/DayPickerSingleDateController.jsx [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+    if (didDateChange) {
+      modifiers = this.deleteModifier(modifiers, this.props.date, 'selected');
+      modifiers = this.addModifier(modifiers, date, 'selected');
+    }
+
+    if (didFocusChange) {
+      values(visibleDays).forEach((days) => {
+        Object.keys(days).forEach((day) => {
+          const momentObj = moment(day);
+          if (isDayBlocked(momentObj)) {
```
Comment (timhwang21) [inline src/components/DayPickerSingleDateController.jsx]: @majapw I think this lost the change fixed in #550. Specifically, checks for `isBlocked` and `isOutsideRange`. Cheers!

Comment (majapw) [inline src/components/DayPickerSingleDateController.jsx]: Whoooops. Man, I knew I should have rebased. :|

Comment (majapw) [inline src/components/DayPickerSingleDateController.jsx]: Okay, I think this is back now and released in 12.2.2. D:

**Pull request merged**